stream_alert/rule_processor/payload.py

```diff
@@ -0,0 +1,328 @@
+                proper `pre_parsed_record` set. Conforming to the interface of
+                returning a generator provides the ability to support multi-record
+                payloads, such as those similar to S3.
+        """
+
+    def refresh_record(self, new_record):
+        """Replace the currently loaded record with a new one.
+
+        Used mainly when S3 is used as a source, due to looping over files
+        downloadd from S3 events versus all records being readily available
+        from a Kinesis stream.
+
+        Args:
+            new_record [string]: A new raw record to be parsed
+        """
+        self.pre_parsed_record = new_record
+        self.type = None
+        self.log_source = None
+        self.records = None
+        self.valid = False
+
+
+class S3ObjectSizeError(Exception):
+    """Exception indicating the S3 object is too large to process"""
+    pass
+
+
+class S3Payload(StreamPayload):
+    """S3Payload class"""
+    s3_object_size = 0
+
+    def service(self):
+        return 's3'
+
+    def pre_parse(self):
+        """Pre-parsing method for S3 objects that will download the s3 object,
+        open it for reading and iterate over lines (records) in the file.
+        This yields back references of this S3Payload instance to the caller
+        with a propertly set `pre_parsed_record` for this record.
+
+        Returns:
+            [generator] Yields instances of `self` back to the caller with the
+                proper `pre_parsed_record` set. Conforms to the interface of
+                returning a generator, providing the ability to support
+                multi-record like this (s3).
+        """
+        s3_file = self._get_object()
+        line_num, processed_size = 0, 0
+        for line_num, data in S3Payload._read_s3_file(s3_file):
+
+            self.refresh_record(data)
+            yield self
+
+            # Only do the extra calculations below if debug logging is enabled
+            if not LOGGER.isEnabledFor(log_level_debug):
+                continue
+
+            # Add the current data to the total processed size, +1 to account for line
+            # feed
+            processed_size += (len(data) + 1)
+
+            # Log a debug message on every 100 lines processed
+            if line_num and line_num % 100 == 0:
+                avg_record_size = ((processed_size - 1) / line_num)
+                if avg_record_size:
+                    approx_record_count = self.s3_object_size / avg_record_size
+                    LOGGER.debug(
+                        'Processed %s records out of an approximate total of %s '
+                        '(average record size: %s bytes, total size: %s bytes)',
+                        line_num,
+                        approx_record_count,
+                        avg_record_size,
+                        self.s3_object_size)
+
+    def _download_object(self, region, bucket, key):
```

Comment:
This is within the s3payload class now and is implicit that it's downloading an s3 object - going to leave it as is